### PR TITLE
[Resolve #283] Add Support for Multiple --var-file Options

### LINKED ISSUE #283: Add Support for Multiple `--var-file` Options
Currently, If I supply multiple `--var-file` options:

`sceptre --var-file=a.yaml --var-file=b.yaml --var-file=c.yaml launch-env api`

Sceptre will silently ignore `a.yaml` and `b.yaml`, while reading in `c.yaml`. This seems like an unnecessary limitation, especially when you consider that Sceptre already supports specifying multiple `--var` options. 

I believe there are quite a few reasons why you would want to do this. Our specific use case is that we would like one set of var-files to represent AWS account specific configs (dev.yaml, sit.yaml, prod.yaml), while also having a set of var-files which characterize how much the sceptre env will cost (tiny.yaml, stress_test.yaml, full.yaml).

One implementation detail to consider is what to do if the same key exists in multiple files. I think an appropriate solution would use the order in which the `--var-file`s were passed in to determine precedence.

### PR DESCRIPTION
**Pull request by cboss24**
Now able to specify multiple --var-file options when running
sceptre from the CLI. The dictionaries from these files will
be merged and exposed to the jinja templating. This allows
users to break a monolithic variable file into composable
pieces.

**Replace the text above the line with a summary of the changes in your PR.
The more detailed you are, the better. Before submitting the PR, make sure the following are checked.**

-----------------

* [x] Wrote [good commit messages][1].
* [ ] [Squashed related commits together][2] after the changes have been approved.
* [x] Commit message starts with `[Resolve #issue-number]` (if a related issue exists).
* [x] Added unit tests.
* [ ] Added integration tests (if applicable).
* [x] All unit tests (`make test`) are passing.
* [x] Used the same coding conventions as the rest of the project.
* [x] The new code doesn't generate flake8 (`make lint`) offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

### CODE REVIEW

File: tests/test_cli.py [hunk 1]
```diff
@@ -53,6 +55,48 @@ def raises_exception():
...
+                yaml.safe_dump({"a": 1}, foo)
+                yaml.safe_dump({"b": 2}, bar)
+
+            result = self.runner.invoke(
+                cli,
+                ["--var-file", "foo.yaml", "--var-file", "bar.yaml", "noop"]
+            )
+        self.patcher_getcwd.start()
+
+        templating_vars = yaml.safe_load(result.output)
```
Comment (cboss24) [inline tests/test_cli.py]: this is nasty, I am echoing a dict in the noop command, and then parsing the string output back into a dict here. Open to any suggestions on how to test what the cli group is setting in the ctx object.

**Comment by theseanything** [comment]
Sorry for the late input - I put a implementation comment in the issue thread.

**Comment by cboss24** [comment]
okay no longer throwing the error, now just producing a log.

**Comment by cboss24** [comment]
also rebased off master, solution is now using the new CLI format

File: sceptre/cli/__init__.py [hunk 2]
```diff
@@ -63,7 +63,17 @@ def cli(
...
     if var_file:
-        user_variables.update(yaml.safe_load(var_file.read()))
+        for fh in var_file:
+            parsed = yaml.safe_load(fh.read())
+            # intersection
+            overloaded_keys = set(user_variables.keys()) & set(parsed.keys())
+            if overloaded_keys:
+                logger.debug(
+                    "Duplicate variables encountered: {0}. "
+                    "Using values from file with highest precedence: '{1}'."
```
Comment (theseanything) [inline sceptre/cli/__init__.py]: Not sure we need `Using values from file with highest precedence: ` as it makes the message overly verbose and is pretty easy to which file is taking precedence.

File: sceptre/cli/__init__.py [hunk 3]
```diff
@@ -63,7 +63,17 @@ def cli(
...
+                    .format(', '.join(overloaded_keys), fh.name)
+                )
+            user_variables.update(parsed)
     if var:
         # --var options overwrite --var-file options
```
Comment (theseanything) [inline sceptre/cli/__init__.py]: and we should also probably put a debug with similar effect here as well

File: tests/test_cli.py [hunk 4]
```diff
@@ -55,6 +56,127 @@ def raises_exception():
         raises_exception()
         mock_exit.assert_called_once_with(1)
 
+    def test_single_var(self):
```
Comment (theseanything) [inline tests/test_cli.py]: These test could probably be parametrized (@pytest.mark.parametrize) - there is an example on line 172.

File: sceptre/cli/__init__.py [hunk 5]
```diff
@@ -40,18 +40,18 @@
...
     "--var", multiple=True, help="A variable to template into config files.")
 @click.option(
-    "--var-file", type=click.File("rb"),
+    "--var-file", multiple=True, type=click.File("rb"),
     help="A YAML file of variables to template into config files.")
 @click.pass_context
 @catch_exceptions
 def cli(
         ctx, debug, directory, no_colour, output, var, var_file
-):  # pragma: no cover
```
Comment (theseanything) [inline sceptre/cli/__init__.py]: why did you remove `# pragma: no cover `?

File: sceptre/cli/__init__.py (hunk 2 above)
Comment (b-t-g) [reply sceptre/cli/__init__.py]: Is the verbosity that much of an issue? I suspect it may be more helpful when there were unexpectedly repeated variables (the reasoning is similar to why I advocated for erroring out for duplicates) and it informs the user how we're selecting precedence without having to point them to a guide, which is a decently big usability gain. Especially since this is debug logging which is already verbose, anyway, adding all relevant information seems sensible to me.

File: sceptre/cli/__init__.py (hunk 5 above)
Comment (cboss24) [reply sceptre/cli/__init__.py]: The `cli` function now has full coverage. I thought it would be okay to include it in the coverage report now. That way we will be able to detect if untested paths are added or tests are changed.

File: sceptre/cli/__init__.py (hunk 2 above)
Comment (cboss24) [reply sceptre/cli/__init__.py]: Could we convey the same information with a simpler sentence?

`"Duplicate variables encountered: {}. Using values from {}".format(', '.join(overloaded_keys), fh.name)`

File: sceptre/cli/__init__.py (hunk 3 above)
Comment (cboss24) [reply sceptre/cli/__init__.py]: done.

File: tests/test_cli.py (hunk 4 above)
Comment (cboss24) [reply tests/test_cli.py]: Okay I gave it a go. Although the parametrized solution is less LOC, i'm not sure it's easier to read.

File: sceptre/cli/__init__.py (hunk 2 above)
Comment (theseanything) [reply sceptre/cli/__init__.py]: After this change I'm happy with it!

Comment (cboss24) [reply sceptre/cli/__init__.py]: done!

**Comment by cboss24** [comment]
@seanrankine all done and squashed, let me know if there's anything else!
